### ISSUE TITLE
false-positive(typescript): cannot set default generic parameter value in function definition

### ISSUE THREAD

**Issue by vegerot** [quick-lint/quick-lint-js#921]
what I'm trying to do: create a function that takes a generic `T` input but defaults it to `unknown` if not provided or inferred

```ts
const ident = <T=unknown>(e: T) => e;
```

Expected: 👌 

actual: 
```sh
❯ quick-lint-js --language=experimental-typescript b.ts
b.ts:2:15: error: React/JSX is not allowed in TypeScript code [E0306]
b.ts:2:17: error: unexpected token [E0054]
```

**Comment by vegerot** [comment]
I suspect this is very similar to #919 so I debated even making a separate issue vs. adding a comment to #919.  I can close this issue and add my above post to a comment on #919 if you'd rather

**Comment by vegerot** [comment]
off-topic: quick-lint stops parsing after the first error which is unfortunate

<img width="916" alt="image" src="https://user-images.githubusercontent.com/5297725/214977342-95bd32cc-420b-4e4c-9ee2-40d335a654a9.png">

**Comment by strager** [comment]
> off-topic: quick-lint _partially_ stops parsing after the first error which is unfortunate

This only applies to E0054 (unexpected token), where quick-lint-js gets totally confused and just gives up.

**Comment by strager** [comment]
Do you know what the `=unknown` syntax is called? I don't see any mention of it in TypeScript's documentation. https://www.typescriptlang.org/docs/handbook/2/generics.html

**Comment by vegerot** [comment]
Yes.  Imagine you have a generic function `function foo<T>(e: T) { ... }`
inside the body of `foo`, `e` has the type `any`.  This is technically incorrect and a more accurate type would be `unknown`.

`function foo<T>(e: T) { ... }` is the same as `function foo<T = any>(e: T) {...}`, and in `function foo<T = unknown>(e: T) {...}` `e` is `unknown`.

The way I used it here just affects the the function author and lets them write their generic function in a more type safe way.  It wouldn't affect the caller because `T` will always be inferred from its parameter so it doesn't matter.

Another use case for `<T = foo>` is when `T` is not inferred from its arguments.  For example

```ts
declare async function fetch<T>(url: URL): Promise<T>;
```

In this case, `fetch("google.com")` is an error, so someone wanted to shut up their type checker may use

```ts
declare async function fetch<T = any>(url: URL): Promise<T>;
```

or a gentleman like myself

```ts
declare async function fetch<T = unknown>(url: URL): Promise<FetchResponse<T>>;
```
.

This is commonly used with `extends`.  For example, taking the previous example, maybe we have

```ts
declare async function fetch<T extends FetchResponse>(url: URL): Promise<T>;
```
in this case, `fetch("google.com")` is still an error.  Instead, we will commonly do

```ts
declare async function fetch<T extends FetchResponse = FetchResponse<unknown>(url: URL): Promise<T>;
```
to let `fetch("google.com")` work

**Comment by vegerot** [comment]
essentially, the analogy to function parameters is: `extends` = `:` and `=` = `=` 😛 

idk why they didn't make the syntax the same, so it'd be 
```ts
declare async function fetch<T: FetchResponse = FetchResponse<unknown>>(url: URL): Promise<T>;
```

**Comment by strager** [comment]
I wasn't asking how to *use* the feature. I just want to know what it's called, and maybe some documented caveats with the feature.

**Comment by strager** [comment]
Looking inside the TypeScript compiler repo, perhaps they are called either 'default type arguments' or 'default type parameters'? (Both names are used in the same file.)

https://github.com/microsoft/TypeScript/blob/f794fbf2f6bbc9e69be38d2888858f801c4a51b4/tests/cases/compiler/functionTypeArgumentArityErrors.ts

Or maybe it's just called the 'default type' of the parameter?

https://github.com/microsoft/TypeScript/blob/f794fbf2f6bbc9e69be38d2888858f801c4a51b4/src/compiler/parser.ts#L3833

https://github.com/microsoft/TypeScript/blob/f794fbf2f6bbc9e69be38d2888858f801c4a51b4/src/compiler/factory/nodeFactory.ts#L1477

**Comment by vegerot** [comment]
> I wasn't asking how to _use_ the feature. I just want to know what it's called, and maybe some documented caveats with the feature.

wow I spent 5 minutes working on that for nothing

**Comment by vegerot** [comment]
imo I'd just call it the 'default type' of the parameter

**Comment by strager** [comment]
Fixed in commit b5a6409ec7370779ea57b365247fd15f8625053a.
